code/hud/hudescort.cpp

```diff
@@ -391,99 +462,77 @@ void hud_escort_init()
 //
 void hud_escort_clear_all(bool clear_flags)
 {
-	int i;
-
-	Num_escort_ships = 0;
-	for ( i = 0; i < Max_escort_ships; i++ ) {
-		if(clear_flags && (Escort_ships[i].objnum >= 0) && (Objects[Escort_ships[i].objnum].type == OBJ_SHIP) && (Objects[Escort_ships[i].objnum].instance >= 0)){
-			Ships[Objects[Escort_ships[i].objnum].instance].flags.remove(Ship::Ship_Flags::Escort);
+	if (clear_flags) {
+		for (auto &es : Escort_ships) {
+			if ( (es.objnum >= 0) && (Objects[es.objnum].type == OBJ_SHIP) && (Objects[es.objnum].instance >= 0) ) {
+				Ships[Objects[es.objnum].instance].flags.remove(Ship::Ship_Flags::Escort);
+			}
 		}
-		Escort_ships[i].obj_signature = -99;
-		Escort_ships[i].np_id = -1;
-		Escort_ships[i].escort_hit_timer = 0;
-		Escort_ships[i].escort_hit_next_flash = 0;
-		Escort_ships[i].escort_show_bright = false;
 	}
+
+	Escort_ships.clear();
 }
 
 // internal helper function for sort.
 // sorts first by priority number and then alphabetically
-int escort_compare_func(const void *e1, const void *e2)
+static bool escort_compare(const escort_info &escort1, const escort_info &escort2)
 {
-	escort_info *escort1, *escort2;
-	int diff;
-	int ret;
-
-	escort1 = (escort_info*) e1;
-	escort2 = (escort_info*) e2;
+	if (MULTI_DOGFIGHT) {
+		int n1 = find_player_id(escort1.np_id);
+		int n2 = find_player_id(escort2.np_id);
 
-	// multiplayer dogfight
-	if(MULTI_DOGFIGHT){
-		int n1, n2;
+		if ( (n1 < 0) || (n2 < 0) || (Net_players[n1].m_player == nullptr) || (Net_players[n2].m_player == nullptr) ) {
+			return false;
+		}
 
-		n1 = find_player_id(escort1->np_id);
-		n2 = find_player_id(escort2->np_id);
-		if((n1 < 0) || (n2 < 0) || (Net_players[n1].m_player == NULL) || (Net_players[n2].m_player == NULL)){
-			ret = 0;
-		} else {
-			// player 1 is higher than player 2
-			if(Net_players[n1].m_player->stats.m_kill_count_ok >= Net_players[n2].m_player->stats.m_kill_count_ok){
-				ret = -1;
-			} else {
-				ret = 1;
-			}
+		if (Net_players[n1].m_player->stats.m_kill_count_ok > Net_players[n2].m_player->stats.m_kill_count_ok) {
+			return true;
 		}
 	} else {
-		diff = escort2->priority - escort1->priority;
+		// in multi, players go to the top
+		if (Game_mode & GM_MULTIPLAYER) {
```

Comment:
Maybe we should limit this behavior to Dogfight mode?  If there is a high number of respawns and a very important objective, like a mission critical ship, I know which one I would rather have on the escort list.

Reply:
Dogfight mode is already handled differently and I tried to leave that special handling alone. This specific change should only affect player ships that are manually added, so my thinking was that those ships were particularly important.

It's kind of iffy though and I'm still not sure if high priority mission objectives should come first or not. Needs testing and further discussion.